vispy/visuals/_scalable_textures.py

```diff
@@ -0,0 +1,441 @@
+            data /= clim[1] - clim[0]
+        if should_cast_to_f32(data.dtype):
+            data = data.astype(np.float32)
+        return data
+
+    def scale_and_set_data(self, data, offset=None, copy=True):
+        """Upload new data to the GPU, scaling if necessary."""
+        self._data_dtype = data.dtype
+
+        clim = self._clim
+        is_auto = isinstance(clim, str) and clim == 'auto'
+        if data.ndim == self._ndim or data.shape[self._ndim] == 1:
+            if is_auto:
+                clim = np.min(data), np.max(data)
+            clim = (np.float32(clim[0]), np.float32(clim[1]))
+            data = self._scale_data_on_cpu(data, clim, copy=copy)
+            data_limits = clim
+        else:
+            data_limits = self._get_default_clims(data)
+            if is_auto:
+                clim = data_limits
+
+        self._clim = clim
+        self._data_limits = data_limits
+        return super().scale_and_set_data(data, offset=offset, copy=copy)
+
+
+class GPUScaledTextureMixin(_ScaledTextureMixin):
+    """Texture class for smarter scaling and internalformat decisions.
+
+    This texture class uses internal formats that are not supported by
+    strict OpenGL 2/ES drivers without additional extensions. By using
+    this texture we upload data to the GPU in a format as close to
+    the original data type as possible (32-bit floats on the CPU are 32-bit
+    floats on the GPU). No normalization/scaling happens on the CPU and
+    all of it happens on the GPU. This should avoid unnecessary data copies
+    as well as provide the highest precision for the final visualization.
+
+    The texture format may either be a GL enum string (ex. 'r32f'), a numpy
+    dtype object (ex. np.float32), or 'auto' which means the texture will
+    try to pick the best format for the provided data. By using 'auto' you
+    also give the texture permission to change formats in the future if
+    new data is provided with a different data type.
+
+
+    This class should only be used internally. For similar features where
+    scaling occurs on the CPU see
+    :class:`vispy.visuals._scalable_textures.CPUScaledTextureMixin`.
+
+    To use this mixin, a subclass should be created to combine this mixin with
+    the texture class being used. Existing subclasses already exist in this
+    module. Note that this class **must** appear first in the subclass's parent
+    classes so that its ``__init__`` method is called instead of the parent
+    Texture class.
+
+    """
+
+    # dtype -> internalformat
+    # 'r' will be replaced (if needed) with rgb or rgba depending on number of bands
+    _texture_dtype_format = {
+        np.float32: 'r32f',
+        np.float64: 'r32f',
+        np.uint8: 'r8',
+        np.uint16: 'r16',
+        # np.uint32: 'r32ui',  # not supported texture format in vispy
+        np.int8: 'r8',
+        np.int16: 'r16',
+        # np.int32: 'r32i',  # not supported texture format in vispy
+    }
```

Comment:
```suggestion
    _texture_dtype_format = {
        np.float32: 'r32f',
        np.float64: 'r32f',
        np.uint8: 'r8',  # uint8 normalized
        np.uint16: 'r16',  # uint16 normalized
        # np.int8: 'r8',  # not supported, there are no signed-integer norm formats
        # np.int16: 'r16',
        # np.uint32: 'r32ui',  # not supported, no normal formats for 32bit ints
        # np.int32: 'r32i',
    }
```

Reply:
So I looked into this. The r8 and r16 formats are also normalized unsigned integer formats: the texture data is an uint, but when sampling you get a float in the shader. There are no such formats for signed integers or 32bit integers.

You suggested converting the clim for these signed integer cases, but that is tricky: the data is interpreted as uint, so as you go from 2, 1, 0, -1, there is a jump from 0 to 65525. We could maybe handle it with additional shader code, but not simply by changing the clim values.

Reply:
So I think we'd need to disable these formats for the GPUScaled texture for now.

We can maybe enable them later by enabling support for real integer tectures, but that would require some fundamental changes how we sample such textures.

Reply:
If we know how the data will appear in the GPU then can't we convert the clim to the equivalent number space before it gets used in the shader (`clim_normalized` in this PR)?

Reply:
No, because the int8 (or int16) binary data is interpreted by the GPU as unsigned and this transformation is non-linear. Basically, values -128..0 map to 0.5..1 and values 0..127 map to 0..0.5.

Reply:
In other words: signed ints use the most significant bit to expres the sign. It's set to 1 for negative numbers.

Reply:
Yeah...you're right. It could be done with a lot of special casing with no real benefit.